### PR TITLE
Kernel extension improvements

### PR DESCRIPTION
**Pull request by LuisM000**
## Description
This PR focus is on refining the `Encamina.Enmarcha.SemanticKernel.Extensions.KernelExtensions` class to enhance functionality and address specific use cases.

### GetKernelPromptAsync:

This method allows users to retrieve the final prompt for a given prompt using the provided arguments.

### GetKernelFunctionUsedTokensFromPromptAsync:

This method facilitates the retrieval of the total number of tokens used in generating a prompt from an inline prompt function.

### Fix in GetMaxTokensFromKernelFunction:

It now considers whether the arguments are of type `OpenAIPromptExecutionSettings` when obtaining the MaxTokens. This method is now used in the `QuestionAnsweringPlugin` to calculate the number of tokens.

## How Has This Been Tested?
Tests have been added in `KernelExtensionsTest` covering different cases.

### CODE REVIEW

**Review by rliberoff — CHANGES_REQUESTED** [review]
Changes according to our talk.

**Review by LuisM000 — COMMENTED** [review]
Fixed!

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs [hunk 1]
```diff
@@ -228,44 +283,34 @@ public static IEnumerable<KernelPlugin> ImportPromptFunctionsFromAssembly(this K
     private static int GetMaxTokensFromKernelFunction(Kernel kernel, KernelFunction function, KernelArguments arguments)
     {
         // Try to use IChatCompletionService as the IAService to retrieve the service settings, but fallback to ITextGenerationService if it's not available.
-        // Once the service settings are retrieved, get the value of the `max_tokens` property from the extension data.
+        // Once the service settings are retrieved, get the value of the MaxTokens property from the MaxTokens property or extension data.
         // If the `max_tokens` property is found, check is a JsonElement and try to get its value as an integer.
         // Finally, if the value is an integer, return it.
-        // In any other case (if the service settings are not found, or the `max_tokens` property is not found, or the value is not an integer), throw an exception.
```
Comment (VictorMuMo) [inline src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:234]: In the comment it is said: "In any other case (if the service settings are not found, or the `max_tokens` property is not found, or the value is not an integer), throw an exception." yet in the method no exception is throwed, maybe the comment is outdated?

**Review by VictorMuMo — CHANGES_REQUESTED** [review]

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs (hunk 1 above)
Comment (rliberoff) [reply src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:234]: Yep. It is outdated.

Comment (LuisM000) [reply src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:234]: Fixed, well spotted!

**Review by rliberoff — DISMISSED** [review]

**Review by VictorMuMo — DISMISSED** [review]

**Review by Hiunkeru — DISMISSED** [review]

File: Directory.Build.props [hunk 2]
```diff
@@ -17,7 +17,7 @@
 
   <PropertyGroup>
     <VersionPrefix>8.1.2</VersionPrefix>
-    <VersionSuffix>preview-12</VersionSuffix>
+    <VersionSuffix>preview-13</VersionSuffix>
```
Comment (fjbelizon) [inline Directory.Build.props]: You need update the preview version

Comment (LuisM000) [reply Directory.Build.props]: Fixed!

**Review by VictorMuMo — APPROVED** [review]

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs [hunk 3]
```diff
@@ -101,17 +100,51 @@ public static async Task<string> GetKernelFunctionPromptAsync(this Kernel kernel
...
+    /// <param name="arguments">The arguments passed to the function.</param>
+    /// <param name="promptTemplateFormat">
+    /// The template format for the prompt. This must be provided if <paramref name="promptTemplateFactory"/> is not <see langword="null"/>. This parameter is optional.
+    /// </param>
+    /// <param name="promptTemplateFactory">
+    /// A <see cref="IPromptTemplateFactory"/> to interpret the prompt and its configuration into a <see cref="IPromptTemplate"/>. This parameter is optional.
+    /// </param>
+    /// <param name="cancellationToken">A cancellation token that can be used to receive notice of cancellation.</param>
+    /// <returns>A string containing the generated prompt.</returns>
+    public static Task<string> GetKernelPromptAsync(this Kernel kernel, string promptTemplate, KernelArguments arguments, string promptTemplateFormat = null, IPromptTemplateFactory promptTemplateFactory = null, CancellationToken cancellationToken = default)
```
Comment (fjbelizon) [inline src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:126]: You have to take into consideration the warnings generated on the new code by the integration process. Please check the following warning: Split this 257 characters long line (which is greater than 200 authorised). (https://rules.sonarsource.com/csharp/RSPEC-103)

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs [hunk 4]
```diff
@@ -157,6 +190,28 @@ public static async Task<int> GetKernelFunctionUsedTokensAsync(this Kernel kerne
...
+    /// <param name="tokenLengthFunction">A function to calculate the length of a string in tokens.</param>
+    /// <param name="promptTemplateFormat">
+    /// The template format for the prompt. This must be provided if <paramref name="promptTemplateFactory"/> is not <see langword="null"/>. This parameter is optional.
+    /// </param>
+    /// <param name="promptTemplateFactory">
+    /// A <see cref="IPromptTemplateFactory"/> to interpret the prompt and its configuration into a <see cref="IPromptTemplate"/>. This parameter is optional.
+    /// </param>
+    /// <param name="cancellationToken">A cancellation token that can be used to receive notice of cancellation.</param>
+    /// <returns>The total number of tokens used, plus the maximum allowed response tokens specified in the function.</returns>
+    public static async Task<int> GetKernelFunctionUsedTokensFromPromptAsync(this Kernel kernel, string promptTemplate, KernelFunction function, KernelArguments arguments, Func<string, int> tokenLengthFunction, string promptTemplateFormat = null, IPromptTemplateFactory promptTemplateFactory = null, CancellationToken cancellationToken = default)
```
Comment (fjbelizon) [inline src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:209]: You have to take into consideration the warnings generated on the new code by the integration process. Please check the following warning: Split this 346 characters long line (which is greater than 200 authorized). (https://rules.sonarsource.com/csharp/RSPEC-103)

**Review by fjbelizon — CHANGES_REQUESTED** [review]
You have to take into consideration the warnings generated by your new code in the integration process.

**Comment by LuisM000** [comment]
> You have to take into consideration the warnings generated by your new code in the integration process.

Should I only consider the warnings from my code or all of them (boy scouts rule😀)?

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs (hunk 3 above)
Comment (rliberoff) [reply src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:126]: So, our ruleset in SonarCloud is a bit outdated. This warnings related to 200 characters per line is from times before wide screens, which almost all of us have. This one is safe to be ignored.

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs (hunk 4 above)
Comment (rliberoff) [reply src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:209]: Idem.

So, our ruleset in SonarCloud is a bit outdated. This warnings related to 200 characters per line is from times before wide screens, which almost all of us have. This one is safe to be ignored.

**Comment by rliberoff** [comment]
> > You have to take into consideration the warnings generated by your new code in the integration process.
> 
> Should I only consider the warnings from my code or all of them (boy scouts rule😀)?

Generally, I would say yes. 

But some SonarCloud warnings are outdated and unhelpful. For example, the 200 characters limit per line makes the code harder to read vertically.

**Review by rliberoff — APPROVED** [review]

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs (hunk 3 above)
Comment (fjbelizon) [reply src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:126]: Perfect, I mark it as solved. We should update the ruleset.

File: src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs (hunk 4 above)
Comment (fjbelizon) [reply src/Encamina.Enmarcha.SemanticKernel/Extensions/KernelExtensions.cs:209]: Perfect, I mark it as solved. We should update the ruleset.

**Review by fjbelizon — APPROVED** [review]

**Pull request merged**